apps/ai-content-generator/src/components/app/sidebar/feature-button/FeatureButton.tsx

```diff
@@ -1,31 +1,54 @@
 import { useSDK } from '@contentful/react-apps-toolkit';
 import { SidebarAppSDK } from '@contentful/app-sdk';
-import { Button, Tooltip } from '@contentful/f36-components';
+import { Button, Spinner, Tooltip } from '@contentful/f36-components';
 import featureConfig, { AIFeature } from '@configs/features/featureConfig';
 import { styles } from './FeatureButton.styles';
 import { makeDialogConfig } from '@configs/dialog/dialogConfig';
+import { useState } from 'react';
 
 interface Props {
   feature: AIFeature;
+  isSaving: boolean;
+  setIsSaving: (isSaving: boolean) => void;
 }
 
 interface FieldLocales {
   [key: string]: string[];
 }
 
 const FeatureButton = (props: Props) => {
-  const { feature } = props;
+  const { feature, isSaving, setIsSaving } = props;
   const buttonCopy = featureConfig[feature].buttonTitle;
   const sdk = useSDK<SidebarAppSDK>();
 
+  const [isOpeningDialog, setIsOpeningDialog] = useState(false);
+
   const fields = sdk.entry.fields;
   const fieldLocales = Object.entries(fields).reduce((acc, entryField) => {
     const [fieldId, field] = entryField;
     acc[fieldId] = field.locales;
     return acc;
   }, {} as FieldLocales);
 
-  const handleOnClick = () => {
+  const toggleSaving = (toggleTo: boolean) => {
+    setIsSaving(toggleTo);
+    setIsOpeningDialog(toggleTo);
+  };
+
+  const handleOnClick = async () => {
+    if (isSaving) {
+      return;
+    }
+
+    try {
+      toggleSaving(true);
```

Comment:
toggle isn't representative of what you're doing as you're correctly setting the actual boolean. Maybe `updateSaveStates` is a better name?